## Ex 3.1 :

### Printing number of records .. 

In this exercise you will write a program to explore the sata in the Pregnancies table. 

1. In the directory where you put survey.py and the data files, create file named first.py and type or paste in the following code and check the results: 

        import survey

        table = survey.Pregnancies()

        table.ReadRecords()

        print 'Number of pregnancies: ', len(table.records)
       
As here I'm in julia .. hence, importing survey.py is infeasible .. so, I'm gonna read the data in dataframe table format and print the length of that table .. 

In [1]:
using DataFrames

In [2]:
data = [("caseid",1,12,Int), ("nbrnaliv",22,22,Int),("babysex",56,56,Int), ("birthwgt_lb",57,58,Int),("birthwgt_oz",59,60,Int), ("prglength", 275,276,Int), ("outcome", 277,277, Int), ("birthord",278,279,Int), ("agepreg", 284,287,Int), ("finalwgt",423,440,Float64)]

10-element Array{Tuple{String,Int64,Int64,DataType},1}:
 ("caseid",1,12,Int64)       
 ("nbrnaliv",22,22,Int64)    
 ("babysex",56,56,Int64)     
 ("birthwgt_lb",57,58,Int64) 
 ("birthwgt_oz",59,60,Int64) 
 ("prglength",275,276,Int64) 
 ("outcome",277,277,Int64)   
 ("birthord",278,279,Int64)  
 ("agepreg",284,287,Int64)   
 ("finalwgt",423,440,Float64)

In [3]:
df_=DataFrame()

0×0 DataFrames.DataFrame

In [4]:
map(x->df_[Symbol(x[1])]=Vector{x[4]}(0),data)

10-element Array{Array{T,1},1}:
 Int64[]  
 Int64[]  
 Int64[]  
 Int64[]  
 Int64[]  
 Int64[]  
 Int64[]  
 Int64[]  
 Int64[]  
 Float64[]

In [5]:
typeof(data)

Array{Tuple{String,Int64,Int64,DataType},1}

In [6]:
open("2002FemPreg.dat","r") do io
    for l in eachline(io)
        if length(l)>100
            for (i,case) in enumerate(data)
                val2push=
                try 
                    parse(case[4],l[case[2]:case[3]])
                catch 
                    NA
                end
                push!(df_[Symbol(case[1])],val2push)
          end
      end
    end
end

In [7]:
df_

13593×10 DataFrames.DataFrame
│ Row   │ caseid │ nbrnaliv │ babysex │ birthwgt_lb │ birthwgt_oz │ prglength │
├───────┼────────┼──────────┼─────────┼─────────────┼─────────────┼───────────┤
│ 1     │ 1      │ 1        │ 1       │ 8           │ 13          │ 39        │
│ 2     │ 1      │ 1        │ 2       │ 7           │ 14          │ 39        │
│ 3     │ 2      │ 3        │ 1       │ 9           │ 2           │ 39        │
│ 4     │ 2      │ 1        │ 2       │ 7           │ 0           │ 39        │
│ 5     │ 2      │ 1        │ 2       │ 6           │ 3           │ 39        │
│ 6     │ 6      │ 1        │ 1       │ 8           │ 9           │ 38        │
│ 7     │ 6      │ 1        │ 2       │ 9           │ 9           │ 40        │
│ 8     │ 6      │ 1        │ 2       │ 8           │ 6           │ 42        │
│ 9     │ 7      │ 1        │ 1       │ 7           │ 9           │ 39        │
│ 10    │ 7      │ 1        │ 2       │ 6           │ 10          │ 35        │
│ 11    │ 12     │ 1        │ 1       │ 7           │ 13          │ 39        │
⋮
│ 13582 │ 12568  │ 1        │ 1       │ 6           │ 6           │ 39        │
│ 13583 │ 12568  │ NA       │ NA      │ NA          │ NA          │ 3         │
│ 13584 │ 12569  │ NA       │ NA      │ NA          │ NA          │ 17        │
│ 13585 │ 12569  │ 1        │ 2       │ 6           │ 6           │ 34        │
│ 13586 │ 12570  │ NA       │ NA      │ NA          │ NA          │ 9         │
│ 13587 │ 12570  │ NA       │ NA      │ NA          │ NA          │ 9         │
│ 13588 │ 12570  │ NA       │ NA      │ NA          │ NA          │ 9         │
│ 13589 │ 12571  │ 1        │ 1       │ 6           │ 3           │ 39        │
│ 13590 │ 12571  │ NA       │ NA      │ NA          │ NA          │ 6         │
│ 13591 │ 12571  │ NA       │ NA      │ NA          │ NA          │ 5         │
│ 13592 │ 12571  │ 1        │ 1       │ 7           │ 8           │ 39        │
│ 13593 │ 12571  │ 1        │ 1       │ 7           │ 8           │ 39        │

│ Row   │ outcome │ birthord │ agepreg │ finalwgt │
├───────┼─────────┼──────────┼─────────┼──────────┤
│ 1     │ 1       │ 1        │ 3316    │ 6448.27  │
│ 2     │ 1       │ 2        │ 3925    │ 6448.27  │
│ 3     │ 1       │ 1        │ 1433    │ 12999.5  │
│ 4     │ 1       │ 2        │ 1783    │ 12999.5  │
│ 5     │ 1       │ 3        │ 1833    │ 12999.5  │
│ 6     │ 1       │ 1        │ 2700    │ 8874.44  │
│ 7     │ 1       │ 2        │ 2883    │ 8874.44  │
│ 8     │ 1       │ 3        │ 3016    │ 8874.44  │
│ 9     │ 1       │ 1        │ 2808    │ 6911.88  │
│ 10    │ 1       │ 2        │ 3233    │ 6911.88  │
│ 11    │ 1       │ 1        │ 2575    │ 6909.33  │
⋮
│ 13582 │ 1       │ 1        │ 3066    │ 7772.21  │
│ 13583 │ 5       │ NA       │ 3325    │ 7772.21  │
│ 13584 │ 2       │ NA       │ 2366    │ 5075.16  │
│ 13585 │ 1       │ 1        │ 2691    │ 5075.16  │
│ 13586 │ 2       │ NA       │ 2141    │ 11325.0  │
│ 13587 │ 2       │ NA       │ 2241    │ 11325.0  │
│ 13588 │ 2       │ NA       │ 2341    │ 11325.0  │
│ 13589 │ 1       │ 1        │ 1791    │ 6269.2   │
│ 13590 │ 2       │ NA       │ 1850    │ 6269.2   │
│ 13591 │ 2       │ NA       │ 1975    │ 6269.2   │
│ 13592 │ 1       │ 2        │ 2158    │ 6269.2   │
│ 13593 │ 1       │ 3        │ 2158    │ 6269.2   │

In [8]:
size(df_[1])  ## number of records in df_ table 

(13593,)

## Justification:

This results match the python results as well as the documentation results .. 

http://www.icpsr.umich.edu/nsfg6/Controller?displayPage=labelDetails&fileCode=PREG&section=&subSec=8014&srtLabel=611774

## Ex 3.2: 

### count (alive) .. 

Write a loop that iterates table and counts the number of live births. 

Find the documentation of outcome and confirm that your result is consistent with the summary in the documentation.

In [245]:
## we gonna utilize on (Outcome) column .. 
alive_counter = 0

for i in df_[7]
    if(i == 1)
        alive_counter = alive_counter + 1
    end
end
alive_counter

9148

## Comment:

The number of obtained living births matches the python results and this number is already mentioned in the documentation .. 

http://www.icpsr.umich.edu/nsfg6/Controller?displayPage=labelDetails&fileCode=PREG&section=&subSec=8014&srtLabel=611785

## Ex 3.3 :


### Partition the data .. 

Modify the loop to partition the live birth records into two groups, one for first babies and one for the others. 

Again, read the documentation of birthord to see if your results are consistent.

I have to skip non-live births .. before partitioning the living ones .. 


selecting the living will be from (outcome) column .. 

then, birthord column will be utilized in the partitioning process .. 

In [246]:
## outcom column .. that contains alive and non_alive babies
df_[:,7]

13593-element DataArrays.DataArray{Int64,1}:
 1
 1
 1
 1
 1
 1
 1
 1
 1
 1
 1
 1
 1
 ⋮
 1
 5
 2
 1
 2
 2
 2
 1
 2
 2
 1
 1

In [247]:
##birthord column .. that contains first babies and others 
df_[:,8]

13593-element DataArrays.DataArray{Int64,1}:
 1  
 2  
 1  
 2  
 3  
 1  
 2  
 3  
 1  
 2  
 1  
 1  
 2  
 ⋮  
 1  
  NA
  NA
 1  
  NA
  NA
  NA
 1  
  NA
  NA
 2  
 3  

In [73]:
## Preparing columns ..

first_table = DataFrame()

myArrays=map(x->Vector{x[4]}(0),data)

for (i,case) in enumerate(data)
    first_table[Symbol(case[1])]=myArrays[i]
end

In [74]:
others_table = DataFrame()

myArrays_=map(x->Vector{x[4]}(0),data)

for (i,case) in enumerate(data)
    others_table[Symbol(case[1])]=myArrays_[i]
end

In [75]:
for i = 1:size(df_)[1] #scan number of records 
    if(df_[i, 7] == 1) # alive
        if (df_[i, 8] == 1) #first baby
            append!(first_table, df_[i,:])
        else
            append!(others_table, df_[i, :])
        end
    end
end
first_table 
others_table

4735×10 DataFrames.DataFrame
│ Row  │ caseid │ nbrnaliv │ babysex │ birthwgt_lb │ birthwgt_oz │ prglength │
├──────┼────────┼──────────┼─────────┼─────────────┼─────────────┼───────────┤
│ 1    │ 1      │ 1        │ 2       │ 7           │ 14          │ 39        │
│ 2    │ 2      │ 1        │ 2       │ 7           │ 0           │ 39        │
│ 3    │ 2      │ 1        │ 2       │ 6           │ 3           │ 39        │
│ 4    │ 6      │ 1        │ 2       │ 9           │ 9           │ 40        │
│ 5    │ 6      │ 1        │ 2       │ 8           │ 6           │ 42        │
│ 6    │ 7      │ 1        │ 2       │ 6           │ 10          │ 35        │
│ 7    │ 14     │ 1        │ 1       │ 4           │ 0           │ 37        │
│ 8    │ 15     │ 1        │ 2       │ 7           │ 8           │ 33        │
│ 9    │ 21     │ 1        │ 2       │ 8           │ 3           │ 39        │
│ 10   │ 24     │ 1        │ 2       │ 7           │ 6           │ 39        │
│ 11   │ 24     │ 1        │ 1       │ 6           │ 13          │ 39        │
⋮
│ 4724 │ 12544  │ 1        │ 2       │ 7           │ 2           │ 39        │
│ 4725 │ 12544  │ 1        │ 2       │ 6           │ 8           │ 39        │
│ 4726 │ 12546  │ 1        │ 2       │ 6           │ 14          │ 44        │
│ 4727 │ 12547  │ 1        │ 2       │ 7           │ 11          │ 40        │
│ 4728 │ 12547  │ 1        │ 2       │ 7           │ 10          │ 39        │
│ 4729 │ 12556  │ 1        │ 2       │ 6           │ 11          │ 39        │
│ 4730 │ 12556  │ 1        │ 1       │ 6           │ 0           │ 39        │
│ 4731 │ 12556  │ 1        │ 2       │ 5           │ 13          │ 39        │
│ 4732 │ 12561  │ 1        │ 2       │ 6           │ 2           │ 39        │
│ 4733 │ 12566  │ 1        │ 1       │ 7           │ 0           │ 39        │
│ 4734 │ 12571  │ 1        │ 1       │ 7           │ 8           │ 39        │
│ 4735 │ 12571  │ 1        │ 1       │ 7           │ 8           │ 39        │

│ Row  │ outcome │ birthord │ agepreg │ finalwgt │
├──────┼─────────┼──────────┼─────────┼──────────┤
│ 1    │ 1       │ 2        │ 3925    │ 6448.27  │
│ 2    │ 1       │ 2        │ 1783    │ 12999.5  │
│ 3    │ 1       │ 3        │ 1833    │ 12999.5  │
│ 4    │ 1       │ 2        │ 2883    │ 8874.44  │
│ 5    │ 1       │ 3        │ 3016    │ 8874.44  │
│ 6    │ 1       │ 2        │ 3233    │ 6911.88  │
│ 7    │ 1       │ 2        │ 2458    │ 3039.9   │
│ 8    │ 1       │ 2        │ 3033    │ 5553.5   │
│ 9    │ 1       │ 2        │ 3058    │ 7237.12  │
│ 10   │ 1       │ 2        │ 2575    │ 7424.84  │
│ 11   │ 1       │ 3        │ 2941    │ 7424.84  │
⋮
│ 4724 │ 1       │ 2        │ 2116    │ 7334.18  │
│ 4725 │ 1       │ 3        │ 2633    │ 7334.18  │
│ 4726 │ 1       │ 2        │ 2816    │ 9415.21  │
│ 4727 │ 1       │ 2        │ 2108    │ 11499.6  │
│ 4728 │ 1       │ 3        │ 2333    │ 11499.6  │
│ 4729 │ 1       │ 2        │ 2441    │ 3965.7   │
│ 4730 │ 1       │ 3        │ 2783    │ 3965.7   │
│ 4731 │ 1       │ 4        │ 3316    │ 3965.7   │
│ 4732 │ 1       │ 2        │ 2441    │ 4497.3   │
│ 4733 │ 1       │ 2        │ 2591    │ 2627.55  │
│ 4734 │ 1       │ 2        │ 2158    │ 6269.2   │
│ 4735 │ 1       │ 3        │ 2158    │ 6269.2   │

In [77]:
size(first_table)[1], size(others_table)[1]

(4413,4735)

Tables are partitioned ;) ..

The results match python code and the documentation .. 

## Ex 3.4: 

Compute the average pregnancy length (in weeks) for first babies and others. Is there a difference between the groups? How big is it?

In [80]:
mean(first_table[6]), mean(others_table[6])

(38.60095173351461,38.52291446673706)

The averages of first babies and other babies match python code .. 

let's claculate the difference between them in weeks as well as in days .. 

In [83]:
diff_ = mean(first_table[6])- mean(others_table[6])

## difference in days ..
diff_*7

0.5462608674428466

The difference in days matches the python results .. this is an evidance that the first babies don't necessarly come late as the difference between their gesetation and the other's is only about 0.55 days ..